File: backend/queries/images.php
<?php 
require __DIR__ . "/query_base.php";
require $_SERVER["DOCUMENT_ROOT"] . "/models/images.php";

class ImageQueries extends Queries {
    public $table = 'images';

    function OrderImageList($images, $rules=""){
        switch ($rules) {
            case "":
                $images = array_reverse($images);
                return $images;
        }
    }

    function GetAllImages(){
        $this->ConnectDB();
        // unsafe, technically doesnt need to be safe because it isnt sensitive data but still
        $query = "SELECT * FROM $this->table";
        $result = mysqli_query($this->db_con, $query);

        if ($result){
            $all = $result->fetch_all();
        } else {
            throw new Exception(mysqli_error($this->db_con));
        }

        $images = [];
        for ($i = 0; $i < count($all); $i++) {
            $image = new Image($all[$i][1], $all[$i][2], $all[$i][3], $all[$i][4], $all[$i][5]);
            array_push($images, $image);
        }

        $this->DisconnectDB();
        return $images;
    }

    function SearchAllImages($searchQueries){
        $this->db_con = $this->ConnectDB();
        $query = "SELECT * FROM $this->table WHERE id > -1 ";

        // Checks if value has something
        // for mediums and tags checks if they are same as search if they are can be either one
        // UNSAFE NEEDS UPDATE
        foreach ($searchQueries as $key => $value) {
            switch ($key){
                case "search":
                    if ($value == "%%"){ break; }
                    $query .= "AND title LIKE '%$value%' ";
                    break;
                case "mediums":
                    if ($value == "%%"){ break; }
                    if ($value == $searchQueries["search"]) { $query .= "OR "; } else { $query .= "AND "; }
                    $query .= "mediums LIKE '%$value%' "; 
                    break;
                case "tags":
                    if ($value == "%%"){ break; }
                    if ($value == $searchQueries["search"]) { $query .= "OR "; } else { $query .= "AND "; }
                    $query .= "tags LIKE '%$value%' "; 
                    break;
                case "created":
                    if ($value == "%%"){ break; }
                    $query .= "AND created LIKE '%$value%' ";
                    break;
            }
        }

        $result = mysqli_query($this->db_con, $query);

        $all = mysqli_fetch_all($result);
        $images = [];
        for ($i = 0; $i < count($all); $i++) {
            $image = new Image($all[$i][1], $all[$i][2], $all[$i][3], $all[$i][4], $all[$i][5]);
            array_push($images, $image);
        }

        $this->DisconnectDB();

        return $images;
    }

    function PostImage(Image $image){
        $this->ConnectDB();
        $created = date_format($image->created,"Y-m-d");
        //SECURE USE AS EXAMPLE
        $query = "INSERT INTO $this->table (filename, title, tags, mediums, created)
        VALUES (?, ?, ?, ?, ?)";

        $stmt = mysqli_prepare($this->db_con, $query);
        mysqli_stmt_bind_param($stmt, "sssss", $image->filename, $image->title, $image->tags, $image->mediums, $created);

        $result = mysqli_stmt_execute($stmt);
        mysqli_stmt_close($stmt);

        $this->DisconnectDB();

        return $result ? true : false;
    }
}

return new ImageQueries();

File: backend/queries/images.py
import os
import base64
from fnmatch import fnmatch
from enum import Enum

from .base import Query, quick_sort_date
from models.images import *

class OrderStyle(Enum):
    DEFAULT = "default"
    DATE_CREATED = "date_created"

class ImageQuery(Query):
    table = "images"
    image_repo = "./images/"

    def order_image_list(self, images: list, sort_style: OrderStyle) -> list:
        match sort_style:
            case OrderStyle.DEFAULT:
                images.reverse()
            case OrderStyle.DATE_CREATED:
                quick_sort_date(images)
                images.reverse()
        return images
    

    def get_image_file(self, image: ImageOut) -> str | dict:
        local = self.image_repo + image.filename
        if os.path.isfile(local):
            return local
        return {"Error": "Could not find file."}
    

    def upload_image_file(self, image: ImageIn) -> bool:
        extensions = ("jpg", "jpeg", "png")
        base_64 = image.base_64
        split = base_64.split(",")
        file_type, base_64 = split[0].split(";")[0].split("/")[1], split[1]

        if file_type in extensions:
            base_64 = base64.b64decode(base_64)
            count = 0
            for extension in extensions:
                for file in os.listdir(self.image_repo):
                    if fnmatch(file, "*." + extension): count += 1

            filename = f"image_{count}" + ".jpg"
            with open(self.image_repo + filename, "wb") as f:
                f.write(base_64)
            
            return filename
        return False
        

    def get_one_image(self, image_id) -> ImageOut | dict:
        self.connect_db()
        cursor = self.db_con.cursor()

        query = f"""
            SELECT * FROM {self.table} WHERE id = %s
        """
        cursor.execute(query, [image_id])
        res = cursor.fetchone()

        self.disconnect_db()
        if res != None:
            image = ImageOut(
                id=res[0],
                filename=res[1],
                title=res[2],
                mediums=res[3],
                tags=res[4],
                created=res[5],
                description=res[6],
                orientation=res[7]
            )
            return image
        else: 
            return {"Error": f"Image of id {image_id} not found."}


    def get_all_images(self) -> list:
        self.connect_db()
        cursor = self.db_con.cursor()
        
        query = f"""
            SELECT * FROM {self.table}
        """
        cursor.execute(query)
        res = cursor.fetchall()

        self.disconnect_db()

        all_images = []
        for image in res:
            _image = ImageOut(
                id=image[0],
                filename=image[1],
                title=image[2],
                mediums=image[3],
                tags=image[4],
                created=image[5],
                description=image[6],
                orientation=image[7]
            )
            all_images.append(_image)

        return all_images
    

    def get_range_of_images(self, range_start: int, range_end: int, queries = None):
        self.connect_db()
        cursor = self.db_con.cursor()
        query = f"""
            SELECT * FROM {self.table}
            WHERE id >= %s AND id <= %s
        """
        values = [range_start, range_end]

        cursor.execute(query, values)
        res = cursor.fetchall()

        self.disconnect_db()

        all_images = []
        for image in res:
            _image = ImageOut(
                id=image[0],
                filename=image[1],
                title=image[2],
                mediums=image[3],
                tags=image[4],
                created=image[5],
                description=image[6],
                orientation=image[7]
            )
            all_images.append(_image)
        
        return all_images


    def get_some_images(self, start: int, end: int, queries: dict) -> list:
        self.connect_db()
        cursor = self.db_con.cursor()

        query = f"""
            SELECT * FROM {self.table} 
            WHERE id >= {start} AND id <= {end}
        """
        # this is scuffed and probably needs a rework.
        values = []
        for key in queries:
            value = queries[key]
            match key:
                case "title":
                    if value == "": break
                    query += " AND title LIKE %s"
                case "mediums":
                    if value == "":
                        query += " OR"
                        queries["mediums"] = queries["title"]
                        value = queries["title"]
                    else:
                        query += " AND"

                    if queries["mediums"] == "": break
                    query += " mediums LIKE %s"
                case "tags":
                    if value == "":
                        query += " OR"
                        queries["tags"] = queries["title"]
                        value = queries["title"]
                    else:
                        query += " AND"

                    if queries["tags"] == "": break
                    query += " tags LIKE %s"
                case "created":
                    if value == "": break
                    query += " AND created LIKE %s"
                    
            values.append(f'%{value}%')


        cursor.execute(query, values)
        res = cursor.fetchall()
        self.disconnect_db()

        all_images = []
        for image in res:
            _image = ImageOut(
                id=image[0], 
                filename=image[1],
                title=image[2],
                mediums=image[3],
                tags=image[4],
                created=image[5],
                description=image[6],
                orientation=image[7]
            )
            all_images.append(_image)

        return all_images


    def post_image(self, image, filename) -> ImageOut:
        self.connect_db()
        cursor = self.db_con.cursor()

        query = f"""
            INSERT INTO {self.table} (filename, title, tags, mediums, created, descript, orientation)
            VALUES (%s, %s, %s, %s, %s, %s, %s)
        """
        values = (
            filename,
            image.title,
            image.tags,
            image.mediums,
            image.created,
            image.description, 
            image.orientation
        )

        cursor.execute(query, values)
        self.db_con.commit()

        self.disconnect_db()
        return self.get_all_images()